Add a TypeScript pass, and revive package.json pass

aomarks opened the pull request:
This pass updates an element repo to include typings generated by https://github.com/Polymer/gen-typescript-declarations/ and configures an `update-types` NPM script that can be run to re-generate them.

It also revives the existing package.json pass to be much simpler (just the most minimal possible package.json), and to update the gitignore file.

Also:
- Replace custom `node-git` typings with the Definitely Typed version.
- Add custom typings for BowerConfig and NpmConfig.
- Add a `package-lock.json`.
- Bump engines and types to Node 8 so we can use `util.promisify`.

aomarks committed:
Add a TypeScript pass.

This pass updates an element repo to include typings generated by
https://github.com/Polymer/gen-typescript-declarations/ and configures
an "update-types" NPM script that can be run to re-generate them.

Also:

- Replace custom node-git typings with the Definitely Typed version.
- Add a package-lock.json.

TimvdLippe reviewed (commented):
Maybe also update travis to fail if the types were not updated?

aomarks commented:
> Maybe also update travis to fail if the types were not updated?

Good idea. Same idea as your https://github.com/Polymer/polymer/pull/5003. Assuming we are happy with the slight extra time this will add to every Travis run. Looks like we already have a Travis pass, so hopefully we can just add it there.

justinfagnani commented on cleanup-passes/typescript.ts:
```diff
@@ -0,0 +1,109 @@
+  try {
+    packageJson = await fse.readJson(packageJsonPath);
+  } catch {
+    throw new Error(
```
why not make a minimal one here? I think you can get away with just having scripts and devDeps.

aomarks replied:
Ok -- I rewrote the existing package-json pass to create a super minimal one, and added it to this PR. It looks like the previous code there was pretty out of date and is now superseded by modulizer.

justinfagnani commented on cleanup-passes/typescript.ts:
```diff
@@ -0,0 +1,109 @@
+        `Missing or invalid package.json for element: ${element.ghRepo.name}`);
+  }
+
+  if (packageJson.devDependencies === undefined) {
```
You can be a little more declarative if you want with object spread:

```ts
try {
  packageJson = {
    scripts: {},
    devDependencies: {},
    ...await fse.readJson(packageJsonPath);
  };
} ...
```

aomarks replied:
Nice, although since I'm not used to that pattern I think it would have taken me a sec to figure out what was going on.

I've added a new `NpmConfig` type (shared with the new package-json pass, see above) which declares these fields as undefined. TypeScript doesn't infer that they can't be undefined with the suggested snippet, so another reason I think I'll stick with mine.

justinfagnani commented on cleanup-passes/typescript.ts:
```diff
@@ -0,0 +1,109 @@
+      // If all we did was update the package lock, don't bother with the
+      // commit. But do include it if we changed something else.
+      commitFiles.push(filepath);
+    }
```
do we expect other file paths? If not, should we log unexpected ones?

aomarks replied:
Added a console.log to point these out.

justinfagnani commented on cleanup-passes/typescript.ts:
```diff
@@ -0,0 +1,109 @@
+  // We also have to bower install because the generator needs to see the
+  // dependencies.
+  await execFilePromise('bower', ['install'], execOpts);
+  await execFilePromise('npm', ['run', npmScriptName], execOpts);
```
my add a comment to highlight that this actually runs the generator

aomarks replied:
Done.

aomarks committed:
Rewrite package-json pass to create a super minimal package.json.

It looks like previously this was an early attempt at generating
package.jsons for actually publishing. Now that's done by modulizer, so
all we need here is to create a super minimal one. It also updates
.gitignore to ignore node_modules. If any package.json exists, this pass
now does absolutely nothing.

aomarks committed:
PR suggestions.

aomarks commented:
PTAL!

aomarks committed:
Remove extra slashes.

aomarks commented:
I'm waiting on https://github.com/PolymerLabs/gen-typescript-declarations/pull/71 and https://github.com/PolymerLabs/gen-typescript-declarations/pull/72 to be published before I pull the trigger on running this pass.

aomarks committed:
Use 1.0.0 of generator.

aomarks committed:
Don't add extra newlines to gitignore.

aomarks committed:
Updating types depends on bower. And some other tweaks.

aomarks committed:
Bump generator version.

aomarks commented:
I've run this on more than half the element repos. Tracking at https://github.com/Polymer/gen-typescript-declarations/issues/79.

@justinfagnani I think this can be committed now, PTAL?

aomarks committed:
Add Travis script, and bump generator version.

aomarks commented:
> Maybe also update travis to fail if the types were not updated?

This is done.

TimvdLippe commented on cleanup-passes/typescript.ts:
```diff
@@ -0,0 +1,150 @@
+        // Update the types.
+        'npm run update-types && ' +
+        // If there were any changes, this git command will return non-zero.
+        'git diff --exit-code || ' +
```
🎉 🎉 🎉 🎉 

That error echo is really nice. Maybe update the script in polymer/polymer as well? 😄

aomarks replied:
> That error echo is really nice. Maybe update the script in polymer/polymer as well? 😄

Done in https://github.com/Polymer/polymer/pull/5099

justinfagnani commented on cleanup-passes/package-json.ts:
```diff
@@ -12,146 +12,67 @@
+    // npm warns if any of these fields aren't set.
+    description: bowerConfig.description,
+    repository: bowerConfig.repository,
+    license: bowerConfig.license,
```
The bower config license field is not a standard SPDX license identifier that package.json should have. I'd just hard code this as "BSD-3-Clause"

aomarks replied:
Done.

justinfagnani commented on cleanup-passes/package-json.ts:
```diff
@@ -12,146 +12,67 @@
-'use strict';
-
-import * as fs from 'fs';
+import * as fse from 'fs-extra';
```
I think we can just import this as `fs`

aomarks replied:
I much prefer giving it a distinct name, since actually all the APIs I use are *not* part of the standard fs, and I've seen `fse` in some other repos.

aomarks committed:
Commit travis changes.

aomarks committed:
BSD-3-Clause

aomarks committed:
Fix false
